### DISCUSSION TITLE
Broken Dag, unable to load custom python modules on PYTHONPATH

### DISCUSSION THREAD

**Discussion by shane-lab** [apache/airflow#21608] [Q&A] [answered]
I want to extend certain operators through custom python modules, but I am unable to import these properly from within a DAG. My PYTHONPATH points to the correct directories where these modules resides. It seems as if it can only import modules relatively from the root of the DAG folder path defined through the airflow config. E.g. if i put my operators inside of the AIRFLOW__CORE__DAGS_FOLDER I can load them fine, but I don't want airflow to interpret my custom modules as DAGs. Nor is this the way I would like to go forward with my airflow setup 

When entering an interactive shell on the running container and check the list of directories through the python repl, I can see it list my the path to my modules and through the interpreter I can also import said custom modules.

running airflow info command also confirms the pythonpath and strangely enough it can load custom plugins without any issues.

<details>

Airflow info output:
```sh
Apache Airflow
version                | 2.2.3
executor               | CeleryExecutor
task_logging_handler   | airflow.providers.amazon.aws.log.s3_task_handler.S3TaskHandler
sql_alchemy_conn       | --snip--
                       |
dags_folder            | /opt/airflow/s3-sync/dags
plugins_folder         | /opt/airflow/s3-sync/plugins
base_log_folder        | /opt/airflow/logs
remote_base_log_folder | s3://airflow-bucket/logs

...

Paths info
airflow_home    | /opt/airflow
system_path     | /home/airflow/.local/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
python_path     | /home/airflow/.local/bin:/opt/airflow/s3-sync/plugins:/opt/airflow/s3-sync/operators:/opt/airflow/s3-s
                | ync/dags:/opt/airflow:/opt/airflow/$PYTHONPATH:/usr/local/lib/python37.zip:/usr/local/lib/python3.7:/u
                | sr/local/lib/python3.7/lib-dynload:/home/airflow/.local/lib/python3.7/site-packages:/usr/local/lib/pyt
                | hon3.7/site-packages:/opt/airflow/config
airflow_on_path | True
```
</details>

note: I use an aws s3 bucket to sync dags, plugins and operators with my containers running airflow instances through a shared volume mounted to these containers. Everything is contained within a subdirectory (s3-sync) in AIRFLOW_HOME (/opt/airflow).

my folder structure:

```sh
airflow home # /opt/airflow
    -- s3-sync/
        -- dags/
        -- plugins/
        -- operators/
            -- my_company
                __init__.py # empty file
                my_operator.py
```

```py
# my_dag.py

import airflow...
import my_company.my_operator
```
My dag yields the following error: `ModuleNotFoundError: No module named 'my_company'`

Although I am able to import said module through the python repl on the same container
```sh
airflow@airflow-scheduler-87dbc9458-k9kks:/opt/airflow$ ls s3-sync/operators/my_company/ -al
total 8
drwxr-sr-x 2 1000 1000   55 Feb 15 13:51 .
drwxr-sr-x 3 1000 1000   74 Feb 15 13:51 ..
-rw-r--r-- 1 1000 1000   99 Feb 15 13:50 __init__.py
-rw-r--r-- 1 1000 1000 1673 Feb 15 13:50 my_operator.py
airflow@airflow-scheduler-87dbc9458-k9kks:/opt/airflow$ python
Python 3.7.12 (default, Dec 21 2021, 11:54:35)
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.path
['', '/opt/airflow/s3-sync/plugins', '/opt/airflow/s3-sync/operators', '/opt/airflow/s3-sync/dags', '/opt/airflow', '/opt/airflow/$PYTHONPATH', '/usr/local/lib/python37.zip', '/usr/local/lib/python3.7', '/usr/local/lib/python3.7/lib-dynload', '/home/airflow/.local/lib/python3.7/site-packages', '/usr/local/lib/python3.7/site-packages']
>>> import my_company
>>> print(my_company)
<module 'my_company' from '/opt/airflow/s3-sync/operators/my_company/__init__.py'>
>>>
```

Loading plugins from `"$AIRFLOW_HOME/s3-sync/plugins/*"` works as intended, although I have since noticed this is no longer required to load operators into airflow. importing directly from airflow.operators as it was before in airflow 1 does make more sense to me. 
```sh
airflow@airflow-scheduler-87dbc9458-k9kks:/opt/airflow$ airflow plugins
my_plugin loaded
name      | source                              | operator_extra_links
==========+=====================================+=====================
my_plugin | $PLUGINS_FOLDER/my_company/k8s_plugin.py | GHK8sLink()
```
As [described by the documentation](https://airflow.apache.org/docs/apache-airflow/stable/howto/custom-operator.html#custom-operator), one could also use the plugins path to load operators. This unfortunately seems not to be working either. I've tried both in Airflow generated plugins directory and overridden AIRFLOW__CORE__PLUGINS_FOLDER.

What can be the cause of my airflow setup not recognizing the `operators` path on my PYTHONPATH?

**Comment by potiuk** [comment]
1. You can ignore your custom modules with .airflowignore placed at the top of DAG Folder (one thing worth noticing that unlike .gitignore, it uses regexp).
2. What is the mostl likely problem your PYTHONPATH is not set properly for airflow process. It is very likey not running in interactive session and your PYTHONPATH is only set in .bashr or .bash_profile or similar file. You can double check it by running `airlfow info` instead of `airlfow scheduler` in the same way you run airflow now - and redirecting the output to a file.

**Reply by shane-lab** [reply]
Thanks for your ideas. I would prefer not to go for the first route though. I believe it is possible to fix it somehow. At the moment I am running airflow 1 in production due to the amount of customizations there had been done in the past, but most DAGs can be terminated permanently. So now is a good time to start with a clean install of Airfow 2 and I'd rather not have some preventable workaround in this production environment.

About your second point. I run airflow in an AWS EKS cluster and the PYTHONPATH is set through a configmap. I had already included the airflow info output in my original post. As you can see from there it correctly displays the directories which are on the path. When I omit the PYTHONPATH in my helm release (I use a custom helm chart opposed to official one), the custom plugins can't be resolved either. Meaning the airflow instance can read from the environment the correct path but when the dagbag tries to fill one or more dags with said custom python modules, it is unable to find it outside of the assigned dags folder from airflow core configuration.

**Reply by potiuk** [reply]
I do not know how you run your airlfow and what is your custom chart does but are you sure both workers and scheduler have the same PYTHONPATH ? can you print the PYTHOHPATH you get in one of your tasks? (For example using Bash Operator or Python operator and confirm that your PYTHONPATH is correctly propagated to the workers?

**Reply by potiuk** [reply]
And what does your `sys.path` in Python operator show ?

**Reply by potiuk** [reply]
Maybe you use impersonation, and PYTHONPATH is not set for the impersonated user?

**Reply by shane-lab** [reply]
Ah indeed, I didn't think of just running a task with simple python operator to check the imported path. Good point!

I've created a simple DAG using the PythonOperator that imports the sys path at the top and on task level (a new import statement during the python callable, incase there's any isolation that I am not aware of). 

```py
from airflow import DAG
from airflow.operators.python import PythonOperator
from datetime import datetime

import sys

PATHS = sys.path

def print_callable():
  import sys as s
  print("dag level paths:")
  print(PATHS)
  print("task level paths:")
  print(s.path)

with DAG(
  dag_id="example_dag",
  start_date=datetime(2022, 1, 1),
  schedule_interval="* * * * *",
  catchup=False) as dag:

  task1 = PythonOperator(
    task_id="print_task",
    python_callable=print_callable)

  task1
```

The output of the sys path in both instances is the same when I run the Python repl when ssh into the container.
```sh
[2022-02-16, 16:01:04 UTC] {standard_task_runner.py:76} INFO - Running: ['airflow', 'tasks', 'run', 'example_dag', 'print_task', 'manual__2022-02-16T17:01:03.421780+00:00', '--job-id', '10028', '--raw', '--subdir', 'DAGS_FOLDER/airflow2_k8s/example.py', '--cfg-path', '/tmp/tmpk_xuos_0', '--error-file', '/tmp/tmpoh2r_rgo']
[2022-02-16, 16:01:04 UTC] {standard_task_runner.py:77} INFO - Job 10028: Subtask print_task
[2022-02-16, 16:01:04 UTC] {logging_mixin.py:109} INFO - Running <TaskInstance: example_dag.print_task manual__2022-02-16T17:01:03.421780+00:00 [running]> on host airflow2-worker-85d694bfdf-4kkhg
[2022-02-16, 16:01:04 UTC] {taskinstance.py:1426} INFO - Exporting the following env vars:
AIRFLOW_CTX_DAG_OWNER=airflow
AIRFLOW_CTX_DAG_ID=example_dag
AIRFLOW_CTX_TASK_ID=print_task
AIRFLOW_CTX_EXECUTION_DATE=2022-02-16T17:01:03.421780+00:00
AIRFLOW_CTX_DAG_RUN_ID=manual__2022-02-16T17:01:03.421780+00:00
[2022-02-16, 16:01:04 UTC] {logging_mixin.py:109} INFO - dag level paths:
[2022-02-16, 16:01:04 UTC] {logging_mixin.py:109} INFO - ['/home/airflow/.local/bin', '/opt/airflow/s3-sync/plugins', '/opt/airflow/s3-sync/operators', '/opt/airflow/s3-sync/dags', '/opt/airflow', '/opt/airflow/$PYTHONPATH', '/usr/local/lib/python37.zip', '/usr/local/lib/python3.7', '/usr/local/lib/python3.7/lib-dynload', '/home/airflow/.local/lib/python3.7/site-packages', '/usr/local/lib/python3.7/site-packages', '/opt/airflow/config']
[2022-02-16, 16:01:04 UTC] {logging_mixin.py:109} INFO - task level paths:
[2022-02-16, 16:01:04 UTC] {logging_mixin.py:109} INFO - ['/home/airflow/.local/bin', '/opt/airflow/s3-sync/plugins', '/opt/airflow/s3-sync/operators', '/opt/airflow/s3-sync/dags', '/opt/airflow', '/opt/airflow/$PYTHONPATH', '/usr/local/lib/python37.zip', '/usr/local/lib/python3.7', '/usr/local/lib/python3.7/lib-dynload', '/home/airflow/.local/lib/python3.7/site-packages', '/usr/local/lib/python3.7/site-packages', '/opt/airflow/config']
```

Both PYTHONPATH in scheduler and worker are the same (and even webserver). Every airflow instance runs in a separate pod on my cluster. 

```sh
➜ k exec airflow2-scheduler-87dbc9458-k9kks -c scheduler -n airflow-test -it -- bash
airflow@airflow2-scheduler-87dbc9458-k9kks:/opt/airflow$ echo $PYTHONPATH
/opt/airflow/s3-sync/plugins:/opt/airflow/s3-sync/operators:/opt/airflow/s3-sync/dags:/opt/airflow:$PYTHONPATH
airflow@airflow2-scheduler-87dbc9458-k9kks:/opt/airflow$ exit
exit

➜ k exec airflow2-worker-85d694bfdf-4kkhg -c worker -n airflow-test -it -- bash
airflow@airflow2-worker-85d694bfdf-4kkhg:/opt/airflow$ echo $PYTHONPATH
/opt/airflow/s3-sync/plugins:/opt/airflow/s3-sync/operators:/opt/airflow/s3-sync/dags:/opt/airflow:$PYTHONPATH
```

All of these deployments use the same setup; shared volumes, secrets, configmap and entrypoint. only difference is the arguments that I send to the entrypoint. The pods consists of 2 running containers, one to sync data from S3 bucket and the other to run an instance of airflow with their respected task (webserver, worker, flower, etc.). The containers running Airflow instances are run as Airflow user (uid 5000) and the sync uses root user.

```sh
airflow@airflow2-worker-85d694bfdf-4kkhg:/opt/airflow$ whoami && id
airflow
uid=50000(airflow) gid=0(root) groups=0(root),65534(nogroup)
```

**Comment by shane-lab** [comment, marked as answer]
I have done a bit more of testing and interestingly enough it works (within Airflow) when I manually expand my sys.path with the absolute path to my custom module, but **only with a trailing slash**! I've updated my PYTHONPATH to include trailingslashes as wel for the absolute paths to my modules, but those seem to be stripped within the python interpreter. 

```sh
➜ k exec airflow-worker-85d694bfdf-vcl5c -c worker -n airflow-test -it -- bash 
airflow@airflow-worker-85d694bfdf-vcl5c:/opt/airflow$ echo $PYTHONPATH
/opt/airflow/s3-sync/plugins/:/opt/airflow/s3-sync/operators/:/opt/airflow/s3-sync/dags/:/opt/airflow/:$PYTHONPATH
```

```
[2022-02-17, 14:41:49 UTC] {logging_mixin.py:109} INFO - syspaths paths:
[2022-02-17, 14:41:49 UTC] {logging_mixin.py:109} INFO - ['/home/airflow/.local/bin', '/opt/airflow/s3-sync/plugins', '/opt/airflow/s3-sync/operators', '/opt/airflow/s3-sync/dags', '/opt/airflow', '/opt/airflow/$PYTHONPATH', '/usr/local/lib/python37.zip', '/usr/local/lib/python3.7', '/usr/local/lib/python3.7/lib-dynload', '/home/airflow/.local/lib/python3.7/site-packages', '/usr/local/lib/python3.7/site-packages', '/opt/airflow/config']
```

Is there a way to force append a trailing slash on the custom module paths without having to write it in each and every dag manually? I could have my aws s3 sync script do some transformations as well, but would be an ugly workaround..

```sh
function fix_sys_import() { echo -e "import sys\nsys.path.append(....)\n$(cat $1)" > $1 } 
for f in $(find /opt/airflow/s3-sync/dags -name '*.py'); do fix_sys_import $f; done
```

**Reply by shane-lab** [reply]
I went with the workaround in the end, I'll keep it here as a reference for whoever needs an alternative solution. So the solution I came up with was to extend the `sys.path` for all imported DAGs through my AWS S3 bucket sync.

I use a container to fetch the dags, plugins, operators etc. from a bucket and another container to run an airflow instance. They share the same mounted volume. Here I just repeatedly make a call to my bucket and then override all downloaded python files that gets placed inside the dags folder with a custom entrypoint:

```sh
#!/usr/bin/env bash

: "${PYTHONPATH_EXTRA:=""}"

while getopts ":b:d:i:" opt; do
  case $opt in
    b) bucket="$OPTARG"
    ;;
    d) syncFolder="$OPTARG"
    ;;
    i) interval="$OPTARG"
    ;;
    \?) echo "Invalid option -$OPTARG" >&2
    ;;
  esac
done

if [[ -z $bucket || -z $syncFolder || -z $interval ]]; then
  echo 'One or more variables are not set.'
  echo "Usage: [-b s3://my-bucket -d /local/sync/folder/path -i 10]" 1>&2; exit 1;
fi

function fix_sys_import() {
  local file="$1"
  echo -e "import sys\nsys.path.append(\"$PYTHONPATH_EXTRA")\n$(cat $file)" > $file
}

function sync_s3() {
  local source="$1" destination="$2"
  aws s3 sync ${source} ${destination} --region eu-west-1 --exact-timestamps --exclude "*" --include "*.py" --delete && chown -R 1000:1000 ${destination}
  for f in $(find "$destination/dags" -name '*.py'); do fix_sys_import $f; done
}

while true
do
  sync_s3 $bucket $syncFolder
  sleep $interval
done
```

usage:
```sh
$ /bin/bash entrypoint.sh -b {{ $remoteBucket }} -d {{ $syncFolder }} -i 30
```

**Reply by potiuk** [reply]
@shane-lab How do you sync your S3 folders to be accessible to Airlfow? I think the problem might be that the synchronisation mechanism and filesytem used does not play well with Python path.

For example I found very similar problem here: https://github.com/danielfrg/s3contents/issues/56

And the root cause of it might be that even if S3 might sometimes "look like" filesystem with folders, it is not really one. There are no "folders" in S3. S3 is an object storage and it only have "objects". They are sometimes conveniently displayed as folders (for example in s3 explorer) but they are not.

so for example this structiure that you "see":

```
dags/
            -- my_company
                __init__.py # empty file
                my_operator.py
```

are really two objects stored in s3:

* `dags/my_company/__init__.py`
* `dags/my_company/my_operator.py`

So there is really no `dags/my_company` folder in s3. At all. It is virtually created by having the same prefix with "/" in the name.

If so the "REAL" solution will be to change synchronisation mechanism of yours to turn the s3 into "REAL" folders and "REAL" filesystem.

**Reply by shane-lab** [reply]
@potiuk I have a k8s pod with two containers that use a shared volume. One container runs an instance of Airflow (e.g. scheduler), the second container periodically calls a function to download all python files from a S3 bucket. As shown by my own accepted answer (for the time being and really needed to move on), hereafter I have to iterate over all downloaded files to apply the fix for the module import.

There is no issue with corrupted or missing files in either my mounted volumes or the S3 bucket. I can `cat` the result when SSH'd into the pod's container. The problem when importing my custom modules went away when I expanded the `sys.path` list with paths containing trailing slashes to the module base directory. When I add a task with a bashoperator, I can also assert that the file exists `if [[ ! -f "$AIRFLOW_HOME/s3-sync/operators/my_company/__init__.py" ]] && exit 1`. The sync in fact produces the directories containing the necessary files. 

Also wanted to address that the issue is not with DAG files, but rather any 'custom module' on my PYTHON PATH that was not installed through pip. I can put my custom operators in the AIRFLOW__CORE__DAGS_FOLDER (which then also get synced from S3) and it will load without a problem, but that makes a huge mess in the automated CI/CD workflow and other projects to come.

**Reply by potiuk** [reply]
Could you please explain how you sync your files. Airlfow (and Python) expects a "regular fielsystem"). You msee to have something else. Can you explain the technology you use to sync files from S3?

**Reply by shane-lab** [reply]
This is the easiest drawing I could make to kind of bring the k8s namespace in frame, most importantly here is the scheduler. Although all pods are kind of similar where they all sync with S3 as well. Hopefully it clears things up a bit.
![image](https://user-images.githubusercontent.com/8008368/155902637-234dd088-8bfc-4719-a3cf-e622706dcc49.png)
The S3 sync container executes the shell script (as it's entrypoint) from the accepted answer on this discussion. The volume to which the files are synced to is also shared with the container running Airflow scheduler (and the same goes for worker and webserver). This volume is mounted inside `$AIRFLOW_HOME` (its mount path is `"$AIRFLOW_HOME/s3-sync"`) and acts as any ordinary directory.

**Reply by potiuk** [reply]
I think the onlu possibility I see is that your sync container does non-atomic sync and deletes stuff while syncing. Git Sync of airflow for example is designed in the way to provide "atomic" sync. It syncs all the files from Git to a separate directory and then performs atomic symbolic-link swap when done. Something similar should be done in your sync container.

**Reply by shane-lab** [reply]
hm that is not really the problem though. I am unable to load any custom module through the airflow instance without the root directory having a trailing slash at the end (that was not part of the default path), regardless of path pointed to my mounted volume (s3 sync).

The python interpreter is able to pick it up just fine. It might have been something very simple that I have overlooked. I am not an expert in python, I can write code in any language but that doesn't make me great at writing programs with them. Prepending the piece of code, as described above, to resolve the custom module paths is not ideal, but I havent had any issues with it after 2 weeks since I posted that as a solution to my problem.

**Reply by potiuk** [reply]
Hmm. Very strange.

**Reply by m-aciek** [reply]
May it be somehow related? https://github.com/python/cpython/issues/65238
